### ISSUE TITLE
Help: Implement Soft Delete In BeforeHandler

### ISSUE THREAD

**Issue by resident0** [mevdschee/php-crud-api#599]
Hi @mevdschee  I'm having trouble getting soft delete to work and I hope you could help. This is what I've tried:
```php
'customization.beforeHandler'=>function($operation, $tableName, $request, $environment) {
   if (strtolower($operation) === 'delete') {
     $date = createDateString();
     $body = [];
      $path = $request->getRequestTarget();
      $parts = explode('/', $path);
      $endPoints = explode(',', end($parts));
      if (count($endPoints) == 1) {
            $body = (object)array('Deleted'=>$date);
       } else {
        for ($i=0; $i<count($endPoints); $i++) {
          $body[] = (object)array('Deleted'=>$date);
        }
      }
      $server = $_SERVER;
      $server['REQUEST_METHOD'] = 'PUT';
      $server['CONTENT_TYPE'] = 'application/json';
      $server['CONTENT_LENGTH'] = strlen(json_encode($body));
      $newRequest = Nyholm\Psr7\Factory\Psr17Factory::createServerRequest('PUT', $path, $server);
      $stream = Nyholm\Psr7\Factory\Psr17Factory::createStream(json_encode($body));
      $stream->rewind();
      $Uri = Nyholm\Psr7\Factory\Psr17Factory::createUri("{$server['REQUEST_SCHEME']}://{$server['HTTP_HOST']}{$server['PATH_INFO']}");
      $serverRequest = $newRequest->withUri($Uri)->withBody($stream)->withparsedBody($body);
      foreach ($request->getHeaders() as $key => $value) {
        if ($key !== 'Host')
        $serverRequest = $serverRequest->withAddedHeader($key, $value);
      }
      return $serverRequest->withAddedHeader('Content-Type', 'application/json')->withAddedHeader('Content-Length', strlen(json_encode($body)));
   }
}
```

When I submit a DELETE request I get a response of '0', no rows affected. Is there something I'm overlooking?

**Comment by mevdschee** [comment]
>createUri("{$server['REQUEST_SCHEME']}://{$server['HTTP_HOST']}{$server['PATH_INFO']}

Should that not contain the path as well?

**Comment by resident0** [comment]
``` $Uri ``` in the above code produces:
```php
Nyholm\Psr7\Uri Object
(
    [scheme:Nyholm\Psr7\Uri:private] => http
    [userInfo:Nyholm\Psr7\Uri:private] => 
    [host:Nyholm\Psr7\Uri:private] => rjtesting.ddns.net
    [port:Nyholm\Psr7\Uri:private] => 
    [path:Nyholm\Psr7\Uri:private] => /records/tickets/1
    [query:Nyholm\Psr7\Uri:private] => 
    [fragment:Nyholm\Psr7\Uri:private] => 
)
```

If I change it to ``` createUri("{$server['REQUEST_SCHEME']}://{$server['HTTP_HOST']}{$server['SCRIPT_NAME']}{$server['PATH_INFO']}"); ``` the result is

```php
Nyholm\Psr7\Uri Object
(
    [scheme:Nyholm\Psr7\Uri:private] => http
    [userInfo:Nyholm\Psr7\Uri:private] => 
    [host:Nyholm\Psr7\Uri:private] => rjtesting.ddns.net
    [port:Nyholm\Psr7\Uri:private] => 
    [path:Nyholm\Psr7\Uri:private] => /v2/api.php/records/tickets/1
    [query:Nyholm\Psr7\Uri:private] => 
    [fragment:Nyholm\Psr7\Uri:private] => 
)
```
This results in a 404 response.

If I change it to ``` createUri("{$server['REQUEST_SCHEME']}://{$server['HTTP_HOST']}/api.php{$server['PATH_INFO']}"); ``` the result is

```php
Nyholm\Psr7\Uri Object
(
    [scheme:Nyholm\Psr7\Uri:private] => http
    [userInfo:Nyholm\Psr7\Uri:private] => 
    [host:Nyholm\Psr7\Uri:private] => rjtesting.ddns.net
    [port:Nyholm\Psr7\Uri:private] => 
    [path:Nyholm\Psr7\Uri:private] => /api.php/records/tickets/1
    [query:Nyholm\Psr7\Uri:private] => 
    [fragment:Nyholm\Psr7\Uri:private] => 
)
```
Which also results in a 404.

I have an htaccess rewrite rule in place to allow the script to be omitted from the request URI:
```
RewriteEngine On
RewriteCond %{REQUEST_FILENAME} !-d
RewriteCond %{REQUEST_FILENAME} !-f
RewriteRule ^(.*)$ api.php/$1 [L,QSA]
```

**Comment by mevdschee** [comment]
>Which results in 0 rows affected.

0 rows affected does not nessecarily mean failure. It could mean that you are updating data that not changed.

Are you sure it is not working as expected?

**Comment by resident0** [comment]
I've double checked. There is an entry at index 1, the deleted field is null. If a make a PUT call to the index the field is updated and I get a response of 1, but not when I intercept a DELETE call and replace it with my custom PUT.

**Comment by resident0** [comment]
I looked into it further and found that a successful PUT request has the following property:
```php
[stream:Nyholm\Psr7\ServerRequest:private] => Nyholm\Psr7\Stream Object
        (
            [stream:Nyholm\Psr7\Stream:private] => Resource id #34
            [seekable:Nyholm\Psr7\Stream:private] => 1
            [readable:Nyholm\Psr7\Stream:private] => 1
            [writable:Nyholm\Psr7\Stream:private] => 
            [uri:Nyholm\Psr7\Stream:private] => php://input
            [size:Nyholm\Psr7\Stream:private] => 
        )
```
That property in my custom stream looks like this:
```php
[stream:Nyholm\Psr7\ServerRequest:private] => Nyholm\Psr7\Stream Object
        (
            [stream:Nyholm\Psr7\Stream:private] => Resource id #47
            [seekable:Nyholm\Psr7\Stream:private] => 1
            [readable:Nyholm\Psr7\Stream:private] => 1
            [writable:Nyholm\Psr7\Stream:private] => 1
            [uri:Nyholm\Psr7\Stream:private] => php://temp
            [size:Nyholm\Psr7\Stream:private] => 
        )
``` 
Note the difference in the uri property. In the successful PUT request ``` php://input ``` contains an object with the values to be updates. In my custom stream both ``` php://temp ``` and ``` php://input ``` are empty.

**Comment by mevdschee** [comment]
>In my custom stream both php://temp and php://input are empty.

You may have to "rewind()" the stream, so it can be read again.

**Comment by mevdschee** [comment]
And your table has a column named 'Deleted' with a capital D?

**Comment by mevdschee** [comment]
And `createDateString()` could be replaced with `date('Y-m-d H:i:s')` to test.

**Comment by resident0** [comment]
> column named 'Deleted'...

Yes, I double checked my spelling. It wouldn't be the first time I've made that mistake.

> createDateString() could be replace...

I just hard-codded the date and time I started the testing.

**Comment by resident0** [comment]
I figured it out. My problem was with my authorization.columnHandler function. I have a some fields that are read only so I test the method to prevent UPDATE and CREATE. My test was too broad, neglected to test for DELETE and was returning false by default. This apparently breaks the flow and prevents the request from modified to UPDATE. I changed the function to return true on DELETE and got the expected response.

Thank you for your suggestions and I apologize for again taking your time with what turned out to be a configuration issue.
